### PR TITLE
Ws robustness

### PR DESCRIPTION
**Pull request by sam-github**
connected to strongloop-internal/scrum-nodeops#680

### CODE REVIEW

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
@slnode test please

**Comment by sam-github** [comment]
Reassigned to you, @rmg, I think you know this code better, even though its a pre-req for https://github.com/strongloop/strong-central/pull/18

File: lib/ws-channel.js [hunk 1]
```diff
@@ -309,8 +337,12 @@ WebsocketChannel.prototype.notify = function(message) {
 
 
 WebsocketChannel.prototype.request = function(message, callback) {
+  if (this._closed) {
+    setImmediate(callback.bind(null, new Error(this._closed)));
```
Comment (rmg) [inline lib/ws-channel.js]: You handle the lack of a callback a few lines down, shouldn't this have a similar guard? Alos, `setImmediate()` takes arguments as arguments.. so `setImmediate(callback, Error(this._closed));` might be clearer.

File: lib/ws-channel.js [hunk 2]
```diff
@@ -554,15 +586,17 @@ WebsocketChannel.prototype._fatalError = function(err) {
     err = new Error(err);
   debug('detach on fatal error: %s', err.message);
   this._detach();
+  this._closeWebsocket();
+  this._closed = err.message;
```
Comment (rmg) [inline lib/ws-channel.js]: What about storing the actual `Error` here so that we can preserve a stacktrace closer to the origin?

**Comment by rmg** [comment]
One probable bug to address and one suggestion/comment that can be ignored if wanted. Otherwise, LGTM.

File: lib/ws-channel.js (hunk 2 above)
Comment (sam-github) [inline lib/ws-channel.js]: I thought about that, but then there wasn't any code that could use the error if it was stored here (and fatal errors are already emitted), so .... YAGNI.

**Comment by sam-github** [comment]
@slnode test please

File: lib/ws-channel.js (hunk 2 above)
Comment (rmg) [inline lib/ws-channel.js]: I was thinking in particular of the `request()` callback up above, but it's debatable whether it would be useful anyway.

**Pull request merged**